### ISSUE TITLE
Problem: Draft API is declared in precompiled header.

### ISSUE THREAD

**Issue by evoskuil** [zeromq/libzmq#1989]
https://github.com/zeromq/libzmq/blob/master/src/precompiled.hpp

**Comment by somdoron** [comment]
What is the problem? It is also defined in zmq.h

**Comment by evoskuil** [comment]
It seems to me that it would either be an unnecessary duplication of code or a confusing inconsistency. No other part of the API is declared in the pch.

**Comment by bluca** [comment]
That's done by design, just like in zproject/czmq. Else, the build will break when DRAFT is disabled.

**Comment by evoskuil** [comment]
@bluca I don't follow, what precludes these declarations from being placed in the headers of the files for which they are defined and conditionally excluded in those locations?

**Comment by bluca** [comment]
Because then they would be duplicated X times instead of one, and some would most likely be left behind. Much clearer if they are in single well-known place. Also it's to follow on what zproject/czmq do as a model - public header + private header, ifdef + ifndef

**Comment by evoskuil** [comment]
Repetition is resolved by use of one or more headers to contain the declarations. Even if there is no existing appropriate header in which to place the API, it would still seem preferable to create a place for it as opposed to overloading the pch for this purpose. We could easily resolve this by creating a `zmq_draft.h` for internal use, and including it in the pch - just as is done with the `zmq.h`

**Comment by bluca** [comment]
Sure, no problem in having a different internal header. I picked that one because it was already there and included everywhere, so it resulted in less churn.

**Comment by evoskuil** [comment]
Sure, that's what I figured. If it's a temporary thing I wouldn't bother. But it seems like this would be a perpetual process.

**Comment by somdoron** [comment]
draft.h files is good idea. However I still think it is a good idea to make it explicit in the zmq.h file and not by including a file.

**Comment by bluca** [comment]
@somdoron - sorry, what do you mean?

**Comment by evoskuil** [comment]
@somdoron I'm sort of ambivalent about that choice. There is extra work/risk in manually maintaining duplicate copies of the same API, but test coverage and CI should mitigate most of the risk. But a second (optional) public header seems inelegant.

**Comment by hitstergtd** [comment]
I suggest that the Draft and Deprecated API stays in `zmq.h`, clearly sectioned away, so that it is abundantly crystal to past and future library users.

But if it is thought that the above is NOT a good way forward, then I suggest we extract deprecated AND draft API aspects and put them into their own headers, with conditional inclusion in `zmq.h`. If we go with the latter, then calling them `zmq_draft.h` and `zmq_depr{cated}.h` would be one way to go.

**Comment by evoskuil** [comment]
I'll pull the option that seems to have a minimum level of agreement. If in the future there is a desire to eliminate management of duplicate copies, that change can be easily applied at that time.

In applying this simple change I ran into several errors (eg min/max namespace conflicts) due to the ordering of the `windows.hpp` include. This should not have been a problem at all, so it is an indication there is a problem in lack of a sufficiently deterministic partial ordering of the includes. In fact I've seen this problem before in the lib and just hadn't tracked it down.

To resolve this problem (and to remove a lot of unnecessary code), I relied on `precompiled.hpp` always being included first, always loading `platform.hpp` so that `windows.hpp` can be loaded before anything else (which is basically a requirement to avoid the above problems).

So the precompiled header will start as follows, and no hpp/cpp file should include `platform.hpp`, `windows.hpp`, `zmq.h` or `zmq_draft.h` - although they can without harm as long as `precompiled.hpp` is the first include (which was already a requirement).

```cpp
#ifndef __ZMQ_PRECOMPILED_HPP_INCLUDED__
#define __ZMQ_PRECOMPILED_HPP_INCLUDED__

#include "platform.hpp"

// This must be included before any windows headers are compiled.
#if defined ZMQ_HAVE_WINDOWS
#include "windows.hpp"
#endif

// 0MQ definitions and exported functions
#include "../include/zmq.h"

// 0MQ DRAFT definitions and exported functions
#include "zmq_draft.h"
...
```

**Comment by bluca** [comment]
Note that it's only on Windows that precompiled.hpp is always included. More details in the PR.

Besides that, having a separate private zmq_draft.h like you are adding is good for me. It must not be a public header or in the include/ subdirs. If it is, then the symbols will be publicly available, and you can be sure people will use it by mistake and then file issues when ABI/API changes. And more importantly, if public symbols change ABI/API, distribution maintainers MUST bump ABI / change library name. Those are the rules and there's no escaping them - for good reasons :-)

**Comment by evoskuil** [comment]
@bluca ack, thanks. This issue is resolved by #1992.

**Comment by bluca** [comment]
Thanks for fixing this!
